From: Mauro Carvalho Chehab <mchehab@redhat.com>
Date: Tue, 28 Sep 2010 14:34:19 -0400
Subject: [edac] i7300_edac: better initialize page counts
Message-id: <20100928113419.1e2ada84@pedra>
Patchwork-id: 28457
O-Subject: [PATCH RHEL5 26/27] V4L/DVB: i7300_edac: better initialize page counts
Bugzilla: 487428
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

It is still somewhat fake, as the pages may not be on this exact order,
and may even be used in mirror mode, but this is a best guess than the
other random fake values.

Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/edac/i7300_edac.c b/drivers/edac/i7300_edac.c
index fa4c284..3053a3f 100644
--- a/drivers/edac/i7300_edac.c
+++ b/drivers/edac/i7300_edac.c
@@ -611,7 +611,8 @@ static void i7300_enable_error_reporting(struct mem_ctl_info *mci)
 static int decode_mtr(struct i7300_pvt *pvt,
 		      int slot, int ch, int branch,
 		      struct i7300_dimm_info *dinfo,
-		      struct csrow_info *p_csrow)
+		      struct csrow_info *p_csrow,
+		      u32 *last_page)
 {
 	int mtr, ans, addrBits, channel;
 
@@ -658,6 +659,11 @@ static int decode_mtr(struct i7300_pvt *pvt,
 	p_csrow->grain = 8;
 	p_csrow->nr_pages = dinfo->megabytes << 8;
 	p_csrow->mtype = MEM_FB_DDR2;
+	p_csrow->csrow_idx = slot;
+	p_csrow->first_page = *last_page;
+	*last_page += p_csrow->nr_pages;
+	p_csrow->last_page = *last_page;
+	p_csrow->page_mask = 0;
 
 	/*
 	 * The type of error detection actually depends of the
@@ -769,6 +775,7 @@ static int i7300_init_csrows(struct mem_ctl_info *mci)
 	int rc = -ENODEV;
 	int mtr;
 	int ch, branch, slot, channel;
+	u32 last_page = 0;
 
 	pvt = mci->pvt_info;
 
@@ -806,18 +813,11 @@ static int i7300_init_csrows(struct mem_ctl_info *mci)
 				p_csrow = &mci->csrows[slot];
 
 				mtr = decode_mtr(pvt, slot, ch, branch,
-							dinfo, p_csrow);
+						 dinfo, p_csrow, &last_page);
 				/* if no DIMMS on this row, continue */
 				if (!MTR_DIMMS_PRESENT(mtr))
 					continue;
 
-				p_csrow->csrow_idx = slot;
-
-				/* FAKE OUT VALUES, FIXME */
-				p_csrow->first_page = 0 + slot * 20;
-				p_csrow->last_page = 9 + slot * 20;
-				p_csrow->page_mask = 0xfff;
-
 				rc = 0;
 			}
 		}
